docs/r1cs-docs-example.md

```diff
@@ -0,0 +1,530 @@
+#             // commit the secret values
+#             for &v_i in &v {
+#                 builder = builder.commit_witness_bytes(b"v_i", v_i.as_bytes());
+#             }
+#             use rand::thread_rng;
+#             builder.finalize(&mut thread_rng())
+#         };
+#         let v_blinding: Vec<Scalar> = (0..2 * k).map(|_| Scalar::random(&mut rng)).collect();
+# 
+#         // Construct a `ConstraintSystem` instance for the shuffle gadget
+#         let (mut cs, variables, commitments) =
+#             ProverCS::new(&bp_gens, &pc_gens, transcript, v, v_blinding.clone());
+# 
+#         // Allocate variables and add constraints to the constraint system
+#         let (input_vars, output_vars) = variables.split_at(k);
+#         ShuffleProof::gadget(&mut cs, input_vars, output_vars);
+# 
+#         // Generate proof
+#         let proof = cs.prove()?;
+# 
+#         Ok((ShuffleProof(proof), commitments))
+#     }
+# }
+impl ShuffleProof {
+    /// Attempt to verify a `ShuffleProof`.
+    pub fn verify<'a, 'b>(
+        &self,
+        pc_gens: &'b PedersenGens,
+        bp_gens: &'b BulletproofGens,
+        transcript: &'a mut Transcript,
+        commitments: &Vec<CompressedRistretto>,
+    ) -> Result<(), R1CSError> {
+        // Apply a domain separator with the shuffle parameters to the transcript
+        let k = commitments.len() / 2;
+        transcript.commit_bytes(b"dom-sep", b"ShuffleProof");
+        transcript.commit_bytes(b"k", Scalar::from(k as u64).as_bytes());
+
+        // Build a `ConstraintSystem` instance with the public inputs
+        let (mut cs, variables) =
+            VerifierCS::new(&bp_gens, &pc_gens, transcript, commitments.to_vec());
+
+        // Add constraints to the constraint system
+        let (input_vars, output_vars) = variables.split_at(k);
+        ShuffleProof::gadget(&mut cs, input_vars, output_vars);
+
+        // Verify the proof
+        cs.verify(&self.0)
+    }
+}
+```
+
+## Using the `ShuffleProof`
+
+Here, we use the `ShuffleProof` gadget by first constructing the common inputs to the `prove` and `verify` functions: the Pedersen and Bulletproofs generators. Next, the prover makes the other inputs to the `prove` function: the list of scalar inputs, the list of scalar outputs, and the prover transcript. The prover calls the `prove` function, and gets a proof and a list of committed points that represent the commitments to the input and output scalars.
+
+The prover passes the proof and the commitments to the verifier. The verifier then makes the other inputs to the `verify` function: the verifier transcript. Note that the starting transcript state must be the same for the prover and verifer transcript; if they are not, then the prover and verifier will receive different challenge scalars and the proof will not verify correctly. The verifier then calls the `verify` function, and gets back a `Result` representing whether or not the proof verified correctly.
```

Comment:
we could note that the starting transcript state is what provides domain separation?